plugins/cpp/parser/src/cppparser.cpp

```diff
@@ -339,7 +343,86 @@ std::vector<std::string> CppParser::getDependentParsers() const
   return std::vector<std::string>{};
 }
 
-bool CppParser::preparse(bool dry_)
+std::vector<std::vector<std::string>> CppParser::createCleanupOrder()
+{
+  typedef boost::adjacency_list<boost::vecS, boost::vecS,
+    boost::bidirectionalS> Graph;
+  typedef boost::adjacency_list<>::vertex_descriptor Vertex;
+  typedef std::pair<int, int> Edge;
+
+  Graph g;
+  std::vector<Edge> edges;
+  std::map<std::string, Vertex> fileNameToVertex;
+  std::deque<std::pair<std::string, Vertex>> deleteQueue;
+
+  for(const auto& file : _ctx.fileStatus)
+  {
+    fileNameToVertex[file.first] = boost::add_vertex(g);
+  }
+
+  try
+  {
+    util::OdbTransaction{_ctx.db}([&]
+    {
+      for (const auto& item : _ctx.fileStatus)
+      {
+        auto file = _ctx.srcMgr.getFile(item.first);
+
+        auto inclusions = _ctx.db->query<model::CppHeaderInclusion>(
+          odb::query<model::CppHeaderInclusion>::included == file->id);
+
+        for (const auto& inclusion : inclusions)
+        {
+          bool inserted;
+          model::FilePtr includer = inclusion.includer.load();
+          boost::graph_traits<Graph>::edge_descriptor e;
+          boost::tie(e, inserted) = boost::add_edge(
+            fileNameToVertex.at(includer->path),
+            fileNameToVertex.at(file->path), g);
+        }
+      }
+    });
+  }
+  catch (odb::database_exception&)
+  {
+    LOG(fatal) << "[cppparser] Topological ordering failed!";
+    return std::vector<std::vector<std::string>>();
+  }
+
+  if (fileNameToVertex.empty())
+  {
+    LOG(info) << "[cppparser] No changed files to create topological order!";
+    return std::vector<std::vector<std::string>>();
+  }
+
+  std::vector<std::vector<std::string>> order;
+  std::size_t index = 0;
+  while (!fileNameToVertex.empty())
+  {
+    order.resize(order.size() + 1);
+
+    for (const auto& item : fileNameToVertex)
+    {
+      if (boost::in_degree(item.second, g) == 0)
+      {
+        order[index].push_back(item.first);
+      }
+    }
+
+    for (const std::string& path : order[index])
+    {
+      boost::clear_out_edges(fileNameToVertex[path], g);
+      fileNameToVertex.erase(path);
+    }
+
+    ++index;
+  }
+  LOG(debug) << "[cppparser] Topology has " << index << " levels.";
+
+  return order;
+}
+
+void CppParser::markModifiedFiles()
```

Comment:
Why only included files are marked? What about the includers? Shouldn't those be reparsed too?

Reply:
The `markByInclusion` method marks the **includer** files (recursively).
Please clarify if I misunderstood your concern here.